client-apis/encrypt.md

```diff
@@ -113,68 +171,185 @@ This document refers to the output of the key derivation algorithm as the derive
-  - The cipherkey is the derived data key
-  - The plaintext is an empty byte array
+- [Frame Length](../data-format/message-header.md#frame-length): MUST be the value of the frame size determined above.
+
+After serializing the message header body,
+this operation MUST calculate an [authentication tag](../data-format/message-header.md#authentication-tag)
+over the message header body.
+The value of this MUST be the output of the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+specified by the [algorithm suite](../framework/algorithm-suites.md), with the following inputs:
+
+- The AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- The IV has a value of 0.
+- The cipherkey is the derived data key
+- The plaintext is an empty byte array
+
+With the authentication tag calculated, this operation MUST serialize the
+[message header authentication](../data-format/message-header.md#header-authentication)
+with the following specifics.
+
+- [IV](../data-format/message-header.md#iv): MUST have a value of 0,
+  padded to the [IV length](../data-format/message-header.md#iv-length).
+- [Authentication Tag](../data-format/message-header.md#authentication-tag): MUST have the value
+  of the authentication tag calculated above.
+
+Once the entire message header has been authenticated and serialized,
+the serialized message header MAY be released.
+
+The encrypted message outputted by this operation MUST have a message header equal
+to the message header calculated in this step.
+
+If the algorithm suite contains a signature algorithm and
+this operation is streaming the encrypted message output to the caller,
+this operation MUST input the serialized header to the signature algorithm as soon as it is serialized,
+such that the serialized header isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+## Construct the body
+
+The encrypted message outputted by this operation MUST have a message body equal
+to the message body calculated in this step.
+
+While there MAY still be plaintext left to encrypt,
+this operation MUST either wait for more plaintext to become available,
+wait for an end to the plaintext to be indicated,
+or to perform encryption on the available plaintext to construct frames
+which make up the message body.
+
+Frames MUST be constructed sequentially such that the concatenation of their decryption
+in [sequence number](#.../data-format/message-body/md#sequence-number)
+order is equal to the input plaintext which has been encrypted so far.
+
+If more input plaintext MAY become available
+and there are not enough input plaintext bytes available to create a new
+[regular frame](#../data-format/message-body.md#regular-frame),
+then this operation MUST wait until either enough plaintext bytes become available
+or the caller indicates an end to the plaintext.
```

Comment:
How do we indicate an end to the plaintext? I believe in an earlier doc you mentioned `EOF`. Should we specify that here?

Reply:
This varies based on how streaming is done in the specific language. I updated the EOF to mirror this wording, but if we think it would be more clear to to refer to it as EOF (and maybe add it as a definition), I can do that instead.

Reply:
Up to you. It might make sense to give an example of 1 or 2 languages but leave the spec as language-agnostic as possible.